### PR TITLE
Adds pull requests and pull request reviews

### PR DESCRIPTION
**Pull request by ingkebil**
A lot of changes in this PR!

The gist is that we need to document how we are devleoping in terms of release models. This in turn leads to the fact that we need to document how we do pull requests and how we request a pull request review.
This is very much a work in progress! Who wants to sit together and we finish it like that?

Extra resources:
https://hackernoon.com/how-to-give-and-get-better-code-reviews-e011c3cda55e
https://medium.com/icarus-wings/what-to-look-for-in-a-pull-request-review-bf9b196bd4f6
https://softwareengineering.stackexchange.com/questions/334488/code-review-workflow-where-pull-request-author-must-merge

Git flow
https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow
https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow
https://www.atlassian.com/git/tutorials/comparing-workflows

github flow
https://guides.github.com/introduction/flow/
https://help.github.com/en/articles/github-flow#following-the-github-flow
https://githubflow.github.io/

git rolling release
http://mrlunar.github.io/Git-Rolling-Release/

How to view changes made to manual?

```
git clone -b prs https://github.com/Clinical-Genomics/development.git
cd development
conda create -n D_mkdocs
source activate D_mkdocs
pip install mkdocs
mkdocs serve
```

go to your browser: http://127.0.0.1:8000

### CODE REVIEW

File: docs/dev/models.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+We are supporting and recommending two kind of development models, depending on the needs of the project.
+To understand this documentation, knowledge of git and branching is recommended! Why don't you read up on our [git][git] page?
+
+## Rolling release (github flow)
+
+Some key features of github flow are:
+
+- Master is deployable
+- Master is deployed to production
+- Development happens on master
```
Comment (patrikgrenfeldt) [inline docs/dev/models.md]:
```suggestion
- Development happens on feature branches
```

File: docs/publish/prod.md [hunk 2]
```diff
@@ -2,16 +2,13 @@
...
 
-(all scripts are in `~/servers/resources`)
-
-- Create a PR
-- Get the PR reviewed and approved ... but don't merge yet
-    - Get someone to review your code
-      - Is the code adhering to the language standards?
-      - Where is the unit test?
-      - Did any signatures of existing functions or methods change?
+- Create a branch from the repository
```
Comment (patrikgrenfeldt) [inline docs/publish/prod.md]: Should this happen _after_ my code is ready for deployment? I suggest we make the header "Your code is ready for deployment. Well done bucko! Now what do I do?" fit with the bullets after "Create, edit, rename, move, or delete files." etc

**Comment by ingkebil** [comment]
This is very much a work in progress! Who wants to sit together and we finish it like that?

File: docs/dev/models.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+- Master is deployable
+- Master is deployed to production
+- Development happens in a development branch, which in turn has feature branches
+- Requires a validation for each release
+
+When would you use git flow?
+
+- The workflow is great for a release-based software development
+- Due to the overhead of swimlanes, medium to large projects benifit the most from git flow
```
Comment (henrikstranneheim) [inline docs/dev/models.md]: What does swimlanes mean in this context?

File: docs/dev/models.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+# How to push to production
+
+This only applies to rolling release projects.
+
+Once a feature has been code reviewed and approved in GH, get a sign off from a manager involved in production. Announce that you will push this to production.
```
Comment (henrikstranneheim) [inline docs/dev/models.md]: github (GH)

File: docs/dev/models.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+To avoid that getting a no might endanger the progress of your project, do not merge until you get the green light to deploy. Do not deploy until the deploy is announced on TTNs daily standup.
```
Comment (henrikstranneheim) [inline docs/dev/models.md]: This sentence is a bit unclear. Do we really need the first part?

File: docs/git/index.md [hunk 6]
```diff
@@ -40,7 +40,7 @@ If you are working locally (your computer only), you could also add the `--globa
 
 ## Create a new repository
 
-There are several ways how one can create a new local repository
+There are several ways how one can create a new local repository.
```
Comment (henrikstranneheim) [inline docs/git/index.md]: There are several ways how one can create a new local repository. -> There are several ways to create a new local repository.

File: docs/dev/models.md (hunk 3 above)
Comment (ingkebil) [reply docs/dev/models.md]: I've removed this now and instead wrote "Due to the overhead of a development branch". Formulation is up for discussion. My main point with choosing the word "overhead" is that there is a cost in maintaining master and development branches and that you cannot easily push a feature to production from development. I hope you agree

File: docs/dev/models.md (hunk 4 above)
Comment (ingkebil) [reply docs/dev/models.md]: This section was still a braindump from after SWEDAC. I removed this section now. Let's add this somewhere else

File: docs/dev/models.md (hunk 5 above)
Comment (ingkebil) [reply docs/dev/models.md]: Braindump :) Removed this for now. This needs it's own page.

File: docs/publish/prod.md (hunk 2 above)
Comment (ingkebil) [reply docs/publish/prod.md]: Removed this for now.

File: docs/git/index.md (hunk 6 above)
Comment (ingkebil) [reply docs/git/index.md]: Fixed!

File: docs/dev/models.md (hunk 3 above)
Comment (henrikstranneheim) [reply docs/dev/models.md]: I think you should avoid that according to the model. Any changes to master that is not from develop should be from a hot fix based on master and then back merged onto develop. I think it is more problematic in this model when master and develop is not merged for quite some time. I think the text is fine

**Comment by ingkebil** [comment]
I've expanded what gitflow and github workflow are. I'm still toying around with how to incorporate how to perform a pull request with both workflows. Feel free to comment and keep in mind that it still is a work in progress.

**Comment by ingkebil** [comment]
Can someone approve if it looks good?

File: docs/dev/models.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+*We are using the shared repository model.*
+
+## Rolling release (github flow)
+
+![github flow image][githubflow-image]
+
+Some key features of github flow are:
+
+- Master is is production ready
```
Comment (b4ckm4n) [inline docs/dev/models.md]: is is --> is in

File: docs/dev/models.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+- Fast deployment speed
+- Ideal for one version in production.
+- Requires a verification for each release
+
+When would you use github flow?
+
+- The workflow is great for bleeding edge software development
+- Meaning that github flow scales easily
+
+Do you want to [know more][githubflow]?
```
Comment (b4ckm4n) [inline docs/dev/models.md]: Nice starship troopers reference ;)

File: docs/github/pr.md [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+### Test your changes
+
+![deploy][deploy]
+
+Before any code is deployed into production it needs to be tested in a production like environment: stage. To make your life easier, read up on how to make an update script to update your tool for each environment right [here][update scripts].
+
+### Merge your feature
+
+![merge][merge]
```
Comment (b4ckm4n) [inline docs/github/pr.md]: Image missing

File: docs/dev/models.md (hunk 7 above)
Comment (b4ckm4n) [reply docs/dev/models.md]: no wait:
is is --> is

File: docs/dev/models.md (hunk 8 above)
Comment (ingkebil) [reply docs/dev/models.md]: :beetle:

**Pull request merged**